runtime/runtime-params-estimator/src/rocksdb.rs

```diff
@@ -0,0 +1,311 @@
+
+    if db_config.debug_rocksdb {
+        println!("# Cost: {:?}", cost);
+        print_levels_info(&db);
+    }
+
+    drop(db);
+    tmp_dir.close().expect("Could not clean up temp DB");
+
+    // Store generated input data in a file for reproducibility of any results
+    if db_config.pr_data_path.is_none() {
+        let mut stats_file = std::fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .append(true)
+            .create(true)
+            .open("names-to-stats.txt")
+            .unwrap();
+        let stats_num = std::io::BufReader::new(&stats_file).lines().count();
+        let data_dump_path = format!("data_dump_{:<04}.bin", stats_num);
+
+        let mut dump = std::fs::File::create(&data_dump_path).unwrap();
+        dump.write_all(&data).unwrap();
+
+        writeln!(stats_file, "# DATA {} written to {}", stats_num, data_dump_path)
+            .expect("Writing to \"names-to-stats.txt\" failed");
+    }
+
+    cost
+}
+
+/// Sequentially insert a number of generated key-value pairs and flushes
+///
+/// Keys are {"1", "2", ... } starting at `key_offset`
+/// Values are different slices taken from `input_data`
+fn sequential_inserts(
+    inserts: usize,
+    value_size: usize,
+    input_data: &[u8],
+    key_offset: usize,
+    db: &DB,
+    force_compaction: bool,
+    force_flush: bool,
+) {
+    for i in 0..inserts {
+        let key = (key_offset + i).to_string();
+        let start = (i * value_size) % (input_data.len() - value_size);
+        let value = &input_data[start..(start + value_size)];
+        db.put(&key, value).expect("Put failed");
+    }
+    if force_flush {
+        db.flush().expect("Flush failed");
+    }
+    if force_compaction {
+        db.compact_range::<&[u8], &[u8]>(None, None);
+    }
+}
+
+/// Insert a number of generated key-value pairs and flushes
+///
+/// Keys are pseudo-random and deterministic based on the seed
+/// Values are different slices taken from `input_data`
+fn prandom_inserts(
+    inserts: usize,
+    value_size: usize,
+    input_data: &[u8],
+    key_seed: u64,
+    db: &DB,
+    force_compaction: bool,
+    force_flush: bool,
+) {
+    let mut prng: XorShiftRng = rand::SeedableRng::seed_from_u64(key_seed);
+    for i in 0..inserts {
+        let key = prng.gen::<u64>().to_string();
+        let start = (i * value_size) % (input_data.len() - value_size);
+        let value = &input_data[start..(start + value_size)];
+        db.put(&key, value).expect("Put failed");
+    }
+    if force_flush {
+        db.flush().expect("Flush failed");
+    }
+    if force_compaction {
+        db.compact_range::<&[u8], &[u8]>(None, None);
+    }
+}
+
+fn input_data(db_config: &RocksDBTestConfig, data_size: usize) -> Vec<u8> {
+    let mut data = vec![0u8; data_size];
+    if let Some(path) = db_config.pr_data_path.as_ref() {
+        let mut input = std::fs::File::open(path).unwrap();
+        input.read_exact(&mut data).unwrap();
```

Comment:
Do we want to assert if file contains *more* data?

Reply:
Now that I think about it, yes, we should do that.
I've added a check for that, using `input.seek(SeekFrom::End(0))` to read the file size. That's the best way I know on stable